ml-agents/mlagents/trainers/trainer.py

```diff
@@ -170,59 +173,53 @@ def export_model(self) -> None:
-            )
-            step = min(self.get_step, self.get_max_steps)
-            stats_summary = self.stats_reporter.get_stats_summaries(
-                "Environment/Cumulative Reward"
-            )
-            if stats_summary.num > 0:
-                LOGGER.info(
-                    " {}: {}: Step: {}. "
-                    "Time Elapsed: {:0.3f} s "
-                    "Mean "
-                    "Reward: {:0.3f}"
-                    ". Std of Reward: {:0.3f}. {}".format(
-                        self.run_id,
-                        self.brain_name,
-                        step,
-                        delta_train_start,
-                        stats_summary.mean,
-                        stats_summary.std,
-                        is_training,
-                    )
+        is_training = (
+            "Training."
+            if self.is_training and self.get_step <= self.get_max_steps
+            else "Not Training."
+        )
+        step = min(self.get_step, self.get_max_steps)
+        stats_summary = self.stats_reporter.get_stats_summaries(
+            "Environment/Cumulative Reward"
+        )
+        if stats_summary.num > 0:
+            LOGGER.info(
+                " {}: {}: Step: {}. "
+                "Time Elapsed: {:0.3f} s "
+                "Mean "
+                "Reward: {:0.3f}"
+                ". Std of Reward: {:0.3f}. {}".format(
+                    self.run_id,
+                    self.brain_name,
+                    step,
+                    time.time() - self.training_start_time,
+                    stats_summary.mean,
+                    stats_summary.std,
+                    is_training,
                 )
-                set_gauge(f"{self.brain_name}.mean_reward", stats_summary.mean)
-            else:
-                LOGGER.info(
-                    " {}: {}: Step: {}. No episode was completed since last summary. {}".format(
-                        self.run_id, self.brain_name, step, is_training
-                    )
+            )
+            set_gauge(f"{self.brain_name}.mean_reward", stats_summary.mean)
+        else:
+            LOGGER.info(
+                " {}: {}: Step: {}. No episode was completed since last summary. {}".format(
+                    self.run_id, self.brain_name, step, is_training
                 )
-            self.stats_reporter.write_stats(int(step))
+            )
+        self.stats_reporter.write_stats(int(step))
 
     def _process_trajectory(self, trajectory: Trajectory) -> None:
         """
         Takes a trajectory and processes it, putting it into the update buffer.
         Processing involves calculating value and advantage targets for model updating step.
         :param trajectory: The Trajectory tuple containing the steps to be processed.
         """
-        raise UnityTrainerException(
-            "The process_trajectory method was not implemented."
-        )
+        self._increment_step(len(trajectory.steps))
```

Comment:
If you're willing to switch to https://docs.python.org/3/library/abc.html (which isn't a bad idea anyway) I believe that you can mark `Trainer._process_trajectory()` as abstract but still provide the "default" implementation. But I agree that if you expect each implementation to to call `_increment_step()`, you should instead just call it from somewhere else.

Reply:
I'm not sure I agree with this @andrewcoh -- seems important that the base `_process_trajectory` includes whatever functionality is needed to make the other function calls within the base class work.  If we have the `super` calls it also makes it easier to add more logic to the common `_process_trajectory` if needed.

Reply:
That I don't disagree with @harperj. My main thought is it might help a user 'fail faster' if this raises an exception unless it's obvious enough from other things that `process_trajectory` needs to be implemented.

Reply:
Makes sense -- I think @chriselion's suggestion about ABCs (which I didn't see before posting my review) would fix this issue in any case :)

Reply:
Thanks guys, will switch to ABCs and move `_increment_step()` into the concrete implementations.